scio-grpc/src/test/scala/com/spotify/scio/grpc/GrpcDoFnTest.scala

```diff
@@ -0,0 +1,179 @@
+      request: ConcatRequest,
+      responseObserver: StreamObserver[ConcatResponse]
+    ): Unit = {
+      responseObserver.onNext(
+        ConcatResponse
+          .newBuilder()
+          .setResponse(request.getStringOne + request.getStringTwo)
+          .build()
+      )
+
+      responseObserver.onCompleted()
+    }
+
+    // Returns the concatenations of both words in either order
+    override def concatServerStreaming(
+      request: ConcatRequest,
+      responseObserver: StreamObserver[ConcatResponse]
+    ): Unit = {
+      responseObserver.onNext(
+        ConcatResponse
+          .newBuilder()
+          .setResponse(request.getStringOne + request.getStringTwo)
+          .build()
+      )
+
+      responseObserver.onNext(
+        ConcatResponse
+          .newBuilder()
+          .setResponse(request.getStringTwo + request.getStringOne)
+          .build()
+      )
+
+      responseObserver.onCompleted()
+    }
+  }
+
+}
+
+class GrpcDoFnTest extends PipelineSpec with BeforeAndAfterAll {
+
+  import GrpcDoFnTest._
+
+  val server: Server = ServerBuilder
+    .forPort(LocalPort)
+    .addService(new ConcatServiceImpl())
+    .asInstanceOf[ServerBuilder[_]]
+    .build()
+
+  override def beforeAll(): Unit = server.start()
+
+  override def afterAll(): Unit = server.shutdown()
+
+  "GrpcDoFn" should "issue request and propagate response" in {
+    runWithContext { sc =>
+      val result = sc
+        .parallelize(0 to 3)
+        .grpcLookup[ConcatResponse, ConcatServiceFutureStub](
+          () => NettyChannelBuilder.forTarget(ServiceUri).usePlaintext().build(),
+          ConcatServiceGrpc.newFutureStub,
+          2
+        ) { case (client, i) =>
+          val request = ConcatRequest
+            .newBuilder()
+            .setStringOne(i.toString)
+            .setStringTwo(i.toString)
+            .build()
+          client.concat(request)
+        }
+
+      result should containInAnyOrder(
+        Seq[(Int, Try[ConcatResponse])](
+          0 -> Success(ConcatResponse.newBuilder().setResponse("00").build()),
+          1 -> Success(ConcatResponse.newBuilder().setResponse("11").build()),
+          2 -> Success(ConcatResponse.newBuilder().setResponse("22").build()),
+          3 -> Success(ConcatResponse.newBuilder().setResponse("33").build())
+        )
+      )
+    }
+  }
+
+  it should "issue request and propagate streamed responses" in {
+    runWithContext { sc =>
+      val result = sc
+        .parallelize(0 to 3)
+        .grpcLookupStream[ConcatResponse, ConcatServiceStub](
+          () => NettyChannelBuilder.forTarget(ServiceUri).usePlaintext().build(),
+          ConcatServiceGrpc.newStub,
+          2
+        ) { case (client, observer, i) =>
+          val request = ConcatRequest
+            .newBuilder()
+            .setStringOne(i.toString)
+            .setStringTwo((Char.char2int('a') + i).toChar.toString)
+            .build()
+          client.concatServerStreaming(request, observer)
```

Comment:
this is a neat way to map the collection with lookup. still user has to do this last line manually. what about instead providing two functions for building the request and passing the client method? e.g.

```
.grpcLookupStream[ConcatResponse, ConcatServiceStub](
          () => NettyChannelBuilder.forTarget(ServiceUri).usePlaintext().build(),
          ConcatServiceGrpc.newStub,
          2
        )(
           i => ConcatRequest
               .newBuilder()
               .setStringOne(i.toString)
               .setStringTwo((Char.char2int('a') + i).toChar.toString)
               .build(),
           ConcatServiceStub.concatServerStreaming
       )
```

Reply:
In that case, for symetry I would also change the `grpcLookup` with smth like this:
```scala
  def grpcLookup[GrpcRequest <: Message, GrpcResponse <: Message: Coder, Client <: AbstractFutureStub[Client]](
    channelSupplier: () => Channel,
    clientFactory: Channel => Client,
    maxPendingRequests: Int
  )(
  r: Request => GrpcRequest,
  f: Client => GrpcRequest => ListenableFuture[GrpcResponse]
  )
  
 def grpcLookupStream[GrpcRequest <: Message, GrpcResponse <: Message: Coder, Client <: AbstractStub[Client]](
    channelSupplier: () => Channel,
    clientFactory: Channel => Client,
    maxPendingRequests: Int
  )(
  r: Request => GrpcRequest,
  f: Client => (GrpcRequest, StreamObserver[GrpcResponse]) => Unit
  ) 
```
I have the feeling that this signature will be counter intuitive for some users. @clairemcginty @kellen WDYT ?

Reply:
oh I see. is the problem that calling the client method (`concatServerStreaming` in this case) is not trivial to be done by Scio, because it has more than one parameter?

Reply:
what I was thinking with this is that users can keep a list of such pairs, even in a separate file if they want which can help with readability (they can name them, and easy to know where the mappings are) and reusability. but I understand it can be a bit confusing. Maybe we can provide two flavours of the methods

Reply:
Not really, I think the signature `A => B => C`, meaning this expects a function that returns a function may not be well understood by users with are not familiar with FP

Reply:
oh ok, I thought we think FP first in scio

Reply:
neat!